Combined LocalNeuralNet code with interface/implementation separation…

asmaalrawi opened the pull request:
- Relates to #2617 
-  Adds LocalNeuralNet class which builds on TensorOneD and TensorTwoD to do neural net json parsing and feed-forward inference, and is used by the Feed Forward agent
- Addresses #2693 
- Replaces #2618 
- Replaces #2628 
- …, tensors moved to inner classes, and JSON11 constructors replaced with STL-based interface.

alawibaba committed:
Combined LocalNeuralNet code with interface/implementation separation, tensors moved to inner classes, and JSON11 constructors replaced with STL-based interface.

avilcheslopez commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,47 @@
+            ///             as vectors of floats.
+            /// 
+            /// @throws geopm::Exception if dimensions are incompatible.
+            static std::unique_ptr<LocalNeuralNet> make_unique(std::vector<std::pair<std::vector<std::vector<float> >, std::vector<float> > > input);
```
Suggestion: given the complexity of the parameter's type, for readability, use typedefs to simplify its declaration and at the same time describe what it is. For example:

`typedef std::pair<...> Layer // or NNLayer, NeuralNetLayer... (I'm sure you can think of a nicer name)`

or

```
typedef std::vector<std::vector<float> > Weights
typedef std::vector<float> Biases
typedef std::pair<Weights, Biases> Layer
```

This way the method declaration would naturally be more readable:

`static std::unique_ptr<LocalNeuralNet> make_unique(std::vector<Layer> input)`

avilcheslopez commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,47 @@
+        public:
+            /// @brief Returns a unique_ptr to a concrete object
+            ///        constructed using the underlying implementation
+            ///        from an array of doubles of weights and biases.
```
Minor discrepancy: this comment talks about doubles, while the actual code uses floats.

alawibaba replied:
I meant pairs!

avilcheslopez commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
```
2015 - 2023

Same applies to the other source files

avilcheslopez commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,47 @@
+            ///
+            /// @throws geopm::Exception if input dimension is incompatible
+            /// with network.
+            virtual std::vector<float> model(std::vector<float> inp) = 0;
```
Suggestion: use a verb to make the methods functionality clearer. For example: create_model, generate_model, perform_inference, do_inference, ... (I'm sure you can think of a nicer name)

alawibaba replied:
That's a great call. I think I misremembered what this was called in other libraries and went with this. It looks like it's typically called "forward" and overload function invocation to use this method. I think it might make sense for us to do that, too -- principle of least surprise. What do you think?

avilcheslopez commented on src/LocalNeuralNetImp.hpp:
```diff
@@ -0,0 +1,176 @@
+            /// @brief Class to store and perform operations on 1D Tensors,
+            ///        aka vectors, suitable for use in feed-forward neural
+            ///        networks.
+            class TensorOneD
```
Q: is there a reason why you opted to make the tensor classes internal?

If they are really particular to LocalNeuralNetImp, then it makes sense. Otherwise, my concern is that this causes unnecessary back-and-forth conversion between the STL types and the tensor classes every time you want to interact with a LocalNeuralNetImp object.

alawibaba replied:
This is what @cmcantalupo and I (@alawibaba) came up with -- totally happy to have a discussion about this.

avilcheslopez commented on src/LocalNeuralNetImp.hpp:
```diff
@@ -0,0 +1,176 @@
+                    std::vector<TensorOneD> m_mat;
+            };
+
+            std::vector<std::pair<TensorTwoD, LocalNeuralNetImp::TensorOneD> > m_layers;
```
```suggestion
            std::vector<std::pair<TensorTwoD, TensorOneD> > m_layers;
```

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+
+namespace geopm
+{
+    std::unique_ptr<LocalNeuralNet> LocalNeuralNet::make_unique(std::vector<std::pair<std::vector<std::vector<float> >, std::vector<float> > > input) {
```
Nit: new line before opening brace

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+        }
+    }
+
+    std::vector<float>
+    LocalNeuralNetImp::model(std::vector<float> inp_vector)
```
Nit:

```suggestion
    std::vector<float> LocalNeuralNetImp::model(std::vector<float> inp_vector)
```

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+        std::size_t nlayers = input.size();
+
+        m_layers.resize(nlayers);
+        for (std::size_t idx = 0; idx < nlayers; idx++) {
```
Nit:

```suggestion
        for (std::size_t idx = 0; idx < nlayers; ++idx) {
```
Same suggestion applies in other places.

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+    {
+        std::size_t vec_size = m_vec.size();
+        LocalNeuralNetImp::TensorOneD rval(vec_size);
+        for(std::size_t idx = 0; idx < vec_size; idx++) {
```
Nit:

```suggestion
        for (std::size_t idx = 0; idx < vec_size; idx++) {
```

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+        std::size_t vec_size = m_vec.size();
+        LocalNeuralNetImp::TensorOneD rval(vec_size);
+        for(std::size_t idx = 0; idx < vec_size; idx++) {
+            rval[idx] = 1/(1 + expf(-(m_vec.at(idx))));
```
Nit: you've ensured `0 < vec_size < m_vec.size()`, so you could simply use the indexing operator.

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+
+    LocalNeuralNetImp::TensorOneD::TensorOneD(std::vector<float> input)
+    {
+        set_dim(input.size());
+        std::size_t vec_size = m_vec.size();
+
+        for (std::size_t idx = 0; idx < vec_size; ++idx) {
+            m_vec[idx] = input[idx];
+        }
```
You could leverage `std::vector`'s assignment operator

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,271 @@
+        return m_vec.size();
+    }
+
+    LocalNeuralNetImp::TensorOneD::TensorOneD(std::vector<float> input)
```
Put this constructor next to the other constructors

alawibaba commented:
Thanks for the review @avilcheslopez! I've written responses to a couple -- I'll address the others and send an update.

alawibaba committed:
Addressed Alejandro's comments: typedefs to improve readability, changed method names, minor nits.

Co-authored by: Ali H Mohammad <ali.mohammad@intel.com>
Co-authored by: Asma H Al-Rawi <asma.h.al-rawi@intel.com>

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,265 @@
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        LocalNeuralNetImp::TensorOneD tmp = inp;
```
Very minor note: you only need one of these TensoreOneD objects. Is `inp` being copied into another object mainly for formality to clearly separate between the input parameter and the return value? If so, perhaps this isn't necessary since due to the conversion you have to do from std::vector to TensorOneD, you're already doing this. So maybe you can just rename `inp` to `tmp` or a more descriptive name. Anyway, just a suggestion to avoid unnecessarily creating another copy of the object, not to stop the PR.

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,265 @@
+        return *this;
+    }
+
+    float& LocalNeuralNetImp::TensorOneD::operator[] (std::size_t idx)
```
Nit:

```suggestion
    float& LocalNeuralNetImp::TensorOneD::operator[](std::size_t idx)
```

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,265 @@
+        return m_vec.at(idx);
+    }
+
+    float LocalNeuralNetImp::TensorOneD::operator[] (std::size_t idx) const
```
Nit:

```suggestion
    float LocalNeuralNetImp::TensorOneD::operator[](std::size_t idx) const
```

avilcheslopez commented on src/LocalNeuralNet.hpp:
```diff
@@ -0,0 +1,64 @@
+            ///
+            /// @throws geopm::Exception if input dimension is incompatible
+            /// with network.
+            std::vector<float> operator()(std::vector<float> inp) {
```
Cool!

Nit:

```suggestion
            std::vector<float> operator()(std::vector<float> inp)
            {
```

avilcheslopez commented on src/LocalNeuralNet.cpp:
```diff
@@ -0,0 +1,265 @@
+        return m_vec.size();
+    }
+
+    LocalNeuralNetImp::TensorOneD LocalNeuralNetImp::TensorOneD::operator+(const LocalNeuralNetImp::TensorOneD& other)
```
Nit: we're putting the `&` next to the identifier (rather than the type).

Multiple instances of this throughout.

avilcheslopez commented on test/LocalNeuralNetTest.cpp:
```diff
@@ -0,0 +1,116 @@
+
+using geopm::LocalNeuralNetImp;
+
+class LocalNeuralNetTest : public ::testing::Test
+{
+};
```
If a test fixture is not really needed, please get rid of this and use the regular `TEST` macro

avilcheslopez commented on test/LocalNeuralNetTest.cpp:
```diff
@@ -0,0 +1,116 @@
+
+TEST_F(LocalNeuralNetTest, test_bad_layers)
+{
+    EXPECT_THROW(LocalNeuralNetImp (
```
You could use `GEOPM_EXPECT_THROW` for deeper checking or is there a reason why you preferred `EXPECT_THROW`?
